### PR TITLE
Fixes #18376 - Improved Installable Errata search

### PR DESCRIPTION
**Pull request by parthaa**
Check http://projects.theforeman.org/issues/18376#note-3 for more info.

Go to Content -> Errata and select the Installable check box
Prior to this commit the "installable_for_hosts" method would create a
introduce an extra 'inner join' causing the query to perform painfully
slow. This commit fixes that by selectively adding that duplicate join
table only if required.

The installable repo query also had a duplicate join for the katello
errata table. Which got removed in this commit

Also fixed the unit tests to highlight the point that "Installable
Errata" is determined by the repositories bound to that host.
Basically the query says "Find all applicable errata on this host.
Filter this list further by only including those repos that are enabled
on this host"

Fixed the unit tests to something that make more sense.
Installable errata is determined purely by the repos that are bound to
the host. It does not really matter what environment/content view
this host belongs. CVE gives the master list of repos available to the
host of which only a subset are enabled or bound.

### CODE REVIEW

**Comment by parthaa** [comment]
To put this in perspective

Query for this line
```
irb(main):003:0> ::Katello::Erratum.installable_for_hosts([Host.first]).to_sql
```
Before this commit

```sql
SELECT DISTINCT "katello_errata".* 
FROM   "katello_errata" 
       INNER JOIN "katello_content_facet_errata" 
               ON "katello_content_facet_errata"."erratum_id" = 
                  "katello_errata"."id" 
       INNER JOIN "katello_repository_errata" 
               ON "katello_repository_errata"."erratum_id" = 
                  "katello_errata"."id" 
       INNER JOIN "katello_content_facet_errata" 
                  "content_facet_errata_katello_errata_join" 
               ON "content_facet_errata_katello_errata_join"."erratum_id" = 
                  "katello_errata"."id" 
       INNER JOIN "katello_content_facets" 
               ON "katello_content_facets"."id" = 
       "content_facet_errata_katello_errata_join"."content_facet_id" 
       INNER JOIN katello_content_facet_repositories 
               ON katello_content_facet_repositories.content_facet_id = 
                  katello_content_facet_errata.content_facet_id 
       INNER JOIN katello_repository_errata AS host_repo_errata 
               ON host_repo_errata.erratum_id = katello_errata.id 
WHERE  ( katello_content_facet_repositories.repository_id = 
                host_repo_errata.repository_id ) 
       AND "katello_content_facets"."host_id" = 1 
```

Notice "katello_content_facet_errata" is joined twice and "katello_repository_errata" is joined twice. Which made the query invalid and slowed down the performance by  a huge magnitude.

After this commit
```sql
SELECT DISTINCT "katello_errata".* 
FROM            "katello_errata" 
INNER JOIN      "katello_content_facet_errata" 
ON              "katello_content_facet_errata"."erratum_id" = "katello_errata"."id" 
INNER JOIN      "katello_content_facets" 
ON              "katello_content_facets"."id" = "katello_content_facet_errata"."content_facet_id" 
INNER JOIN      katello_content_facet_repositories 
ON              katello_content_facet_repositories.content_fac et_id = katello_content_facet_errata.content_facet_id
INNER JOIN      katello_repository_errata AS host_repo_errata 
ON              host_repo_errata.erratum_id = katello_errata.id 
AND             katello_content_facet_repositories.repository_id = host_repo_errata.repository_id
WHERE           "katello_content_facets"."host_id" = 1

```
No duplications  of table names. Moreover the where clause for
```sql'
WHERE  ( katello_content_facet_repositories.repository_id = 
                host_repo_errata.repository_id ) 
```

became a inner join condition 

```sql
INNER JOIN      katello_repository_errata AS host_repo_errata 
ON              host_repo_errata.erratum_id = katello_errata.id 
AND             katello_content_facet_repositories.repository_id = host_repo_errata.repository_id
```

**Comment by parthaa** [comment]
Similarly
Query for this line
```
irb(main):004:0> ::Katello::Erratum.installable_for_hosts.to_sql
```

Before this commit

```sql

SELECT DISTINCT "katello_errata".* 
FROM   "katello_errata" 
       INNER JOIN "katello_content_facet_errata" 
               ON "katello_content_facet_errata"."erratum_id" = 
                  "katello_errata"."id" 
       INNER JOIN "katello_repository_errata" 
               ON "katello_repository_errata"."erratum_id" = 
                  "katello_errata"."id" 
       INNER JOIN katello_content_facet_repositories 
               ON katello_content_facet_repositories.content_facet_id = 
                  katello_content_facet_errata.content_facet_id 
       INNER JOIN katello_repository_errata AS host_repo_errata 
               ON host_repo_errata.erratum_id = katello_errata.id 
WHERE  ( katello_content_facet_repositories.repository_id = 
                host_repo_errata.repository_id ) 
```
Notice the duplicate joins for katello_repository_errata

After this commit

```sql
SELECT DISTINCT "katello_errata".* 
FROM            "katello_errata" 
INNER JOIN      "katello_content_facet_errata" 
ON              "katello_content_facet_errata"."erratum_id" = "katello_errata"."id" 
INNER JOIN      katello_cont ent_facet_repositories 
ON              katello_content_facet_repositories.content_facet_id = 
                        katello_content_facet_errata.content_facet_id
INNER JOIN      katello_repository_errata AS host_repo_errata 
ON              host_repo_errata.erratum_id = katello_errata.id 
AND             katello_content_facet_repositories.repository_id = 
                         host_repo_errata.repository_id

```

Notice no such duploicates.

File: test/models/erratum_test.rb [hunk 1]
```diff
@@ -168,16 +168,17 @@ def test_installable_for_hosts
       refute_includes errata, @enhancement
     end
 
-    def test_installable_for_hosts_dev_environment
-      errata = Erratum.installable_for_hosts([@host_dev, @host_without_errata])
-      assert_includes errata, @security
-      assert_includes errata, @bugfix
-      refute_includes errata, @enhancement
+    def test_installable_for_hosts_with_no_bound_repos
+      # make sure the @host_dev has no bound repositories
```
Comment (jlsherrill) [inline test/models/erratum_test.rb]: can you fix this comment to say @host ?  Also @host_dev is no longer used, can you remove it from the setup?

**Comment by jlsherrill** [comment]
ACK pending small comment
